src/ir/auth_logic/ast_construction.h

```diff
@@ -0,0 +1,60 @@
+
+/*
+ * Copyright 2022 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// This file contains the datalog IR (DLIR) which makes the translation from
+// this authorization logic into datalog simpler.
+
+#ifndef SRC_IR_AUTH_LOGIC_AST_CONSTRUCTION_H_
+#define SRC_IR_AUTH_LOGIC_AST_CONSTRUCTION_H_
+
+#include "src/ir/datalog/program.h"
+#include "src/ir/auth_logic/ast.h"
+#include "absl/status/statusor.h"
+#include "src/policy/auth_logic/auth_logic_cc_generator_grammar.inc/AuthLogicParser.h"
+
+namespace raksha::ir::auth_logic {
+
+class Parser {
+public:
+
+//Parses an input string
+
+absl::StatusOr<Program> ParseProgram (absl::string_view prog_text);
+
+private:
+
+absl::StatusOr<Program> ConstructProgram (AuthLogicParser::ProgramContext* context);
```

Comment:
A more conventional way to use a visitor to populate the AST from antlr would be to define a subclass of `AuthLogicBaseVisitor` and then override the `visit` methods. In your implementations of the visit methods, when you encounter a child node you would call `child.accept(this)` to cause the visitor to visit the children. Because the accept method is virtual and CPP has single dynamic dispatch for virtual methods, the compiler will call the implementation of the method that depends on which specific subclass the `accept` is being called on. Normally, this would let you cut out the dynamic casts completely, but because the CPP version of the AST no longer follows the grammar structure exactly, I don't think we can fully avoid dynamic casts. 

Here's an example that illustrates how this would normally look. For a moment, imagine the AST class structure follows the gramma structure exactly and we have a VerbPhrase class like in the [rust version](https://github.com/google-research/raksha/blob/main/rust/tools/authorization-logic/src/ast.rs).
```
class AstConstructionVisitor : public AuthLogicBaseVisitor {

virtual antlrcpp::Any visitPrinFact(AuthLogicParser::PrinFactContext *ctx) override {
    Principal principal = ctx->principal()->accept(this);
    VerbPhrase verpbPhrase = ctx->verbphrase()->accept(this);    // there is no VerbPhrase class in ast.h
    return PrinPhrase(principal, verbPhrase);                // there is no PrinPhrase either
}

}
```

Here calling `ctx->verbphrase()->accept(this)` is important and how you would get rid of those dynamic casts. The return type of `verbphrase()` is `VerbphraseContext` and both `PredphraseContext` and `ActsAsPhraseContext` are subclasses of `VerbphraseContext`. We do not know at compile-time which subclass of `VerbphraseContext` we will get, but because `accept` is virtual, the implementation of that method that gets called will be handled by a vtable lookup and be the one implemented for the subclass. The code that antlr generates for the accept methods is this:

```
antlrcpp::Any AuthLogicParser::PredphraseContext::accept(tree::ParseTreeVisitor *visitor) {
  if (auto parserVisitor = dynamic_cast<AuthLogicVisitor*>(visitor))
    return parserVisitor->visitPredphrase(this);
  else
    return visitor->visitChildren(this);
}
```

The first branch is the one that happens in our case, so all this does is call `visitPredPhrase(this)`, and the implementation for `ActsAsPhraseContext` is similar except it calls `visitActsAsPhrase`. Because the implementations of the visits would be the ones you define, this is enough to get rid of the dynamic casts -- the dynamic dispatch of `accept` handles this for you.

Unfortunately, in our case the structure of the `AST` doesn't follow the structure of the grammar so I think we can't sue this directly. The real implementation of `visitPrinFact` would look like this:
```
virtual antlrcpp::Any visitPrinFact(AuthLogicParser::PrinFactContext *ctx) override {
    Principal principal = ctx->principal()->accept(this);
    // We know that we should return either the Attribute or CanActAs kind of base fact, but we
    // don't know which one because we don't know what `ctx->verbphrase()` is. To construct
   // either one of these we need to give them the principal though.

   // This doesn't quite work:
   return ctx->verbphrase->accep(this, principal);
  // because we can't change accept to take a principal because accept is defined by antlr
  // So we likely need to handle this with a dynamic cast.
}
```

Reply:
Another way to fix this other than changing the structure of the AST would be to change the grammar so that it more closely follows our updated AST -- if we are using the same `.g4` file for the rust and CPP versions we would need to make sure the rust tests all still pass, though.

I think AST construction is actually one of the cases where the visitor pattern does not help much because constructing the AST requires a complete description of the AST traversal anway. The visitor pattern mainly helps when the interesting functionality of the visit only happens for a small part of the tree. I think using a listener for AST construction is actually fine because we need to give a complete description of the traversal anyway.

We can also just go on using dynamic casts in some cases for now.

Reply:
I didn't check if `antlrcpp::Any` will actually let you return the various type from `ast.h` (like `BaseFact` for example). If it does, then changing the grammar to match the structure of `ast.h` and extending the `AuthLogicBaseVisitor` might be a good way to solve this.

If `antlrcpp::Any` won't let you return those types, then the visitor really does not help with AST construction and it's probably best to keep going with what you have.

Reply:
I think @harsha-mandadi found out the hard way that the visitors won't work well for our use case. `antrlcpp::Any` interferes with our use of `unique_ptr` in the data structures. `unique_ptr` can only be moved, but the visitor interface expects the values to be copyable, if I understand the issue correctly. @harsha-mandadi, please chime in if I am wrong.

Reply:
OK sounds good :)